src/history/hs_rec.c

```diff
@@ -270,25 +270,26 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
      * of the key we are about to insert.
      */
     if (F_ISSET(cursor, WT_CURSTD_KEY_SET)) {
-        WT_ERR(cursor->get_key(cursor, &hs_btree_id, hs_key, &hs_start_ts, &hs_counter));
-        if (hs_btree_id == btree->id && start_time_point->ts == hs_start_ts &&
+        WT_ERR(cursor->get_key(
+          cursor, &retrieved_key.btree_id, retrieved_key.ds_key, &hs_start_ts, &hs_counter));
+        if (retrieved_key.btree_id == hs_key->btree_id && tw->start_ts == hs_start_ts &&
           hs_counter == counter) {
-            WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+            WT_ERR(__wt_compare(session, NULL, hs_key->ds_key, retrieved_key.ds_key, &cmp));
             WT_ASSERT(session, cmp != 0);
         }
     }
 #endif
     /* The tree structure can change while we try to insert the mod list, retry if that happens. */
-    while ((ret = __hs_insert_record_with_btree_int(session, cursor, btree->id, key, type, hs_value,
-              start_time_point, stop_time_point, counter)) == WT_RESTART) {
+    while ((ret = __hs_insert_record_with_btree_int(
+              session, cursor, hs_key, hs_value, tw, counter)) == WT_RESTART) {
         WT_STAT_CONN_INCR(session, cache_hs_insert_restart);
         WT_STAT_DATA_INCR(session, cache_hs_insert_restart);
     }
 err:
 #ifdef HAVE_DIAGNOSTIC
     __wt_scr_free(session, &existing_val);
 #endif
-    __wt_scr_free(session, &hs_key);
+    __wt_scr_free(session, &retrieved_key.ds_key);
```

Comment:
Generally, if we have some kind of struct that owns the resource (esp if we're expecting to add more members), I'd prefer that it has helper functions like:
```
__wt_hs_key_init(...);
__wt_hs_key_free(...);
```
It's easier and less error-prone than having to inspect the struct members and figure out what members need to be freed and how to free them. Also, you can set members like `type` to an invalid value which can help catch use-after-free.

Reply:
I decided to only keep the time window change.